File: Server/test_py/serialisation.py
from typing import List, Tuple
from sprites import Entity
import struct, pygame
from random import randint

def pack_int32(value: int) -> bytes:
    return struct.pack('!i', value)

def pack_uint16(value: int) -> bytes:
    return struct.pack('!H', value)

def pack_uint8(value: int) -> bytes:
    return struct.pack('!B', value)

def unpack_int32(data: bytes, offset: int = 0) -> Tuple[int, int]:
    value = struct.unpack('!i', data[offset:offset+4])[0]
    return value, offset + 4

def unpack_uint16(data: bytes, offset: int = 0) -> Tuple[int, int]:
    value = struct.unpack('!H', data[offset:offset+2])[0]
    return value, offset + 2
def unpack_uint8(data: bytes, offset: int = 0) -> Tuple[int, int]:
    value = struct.unpack('!B', data[offset:offset+1])[0]
    return value, offset + 1

"""
# send a byte that corresponds to an index and then data like:
0x00 for in server id uint16
0x01 for x which is a int32 i think idk
0x02 for y which is a int32 i think idk
0x03 for state which is a uint8
0x04 for dirrection which is a uint8
...eventually add extra fields here
"""

# doesn't add in server id
def serialise(e: Entity) -> bytes:
    data = bytearray()

    # print(f"{e.last_state['x']}:{e.pos.x}")
    if e.last_state.get('x') != e.pos.x:
        # print("Added x")
        data.append(0x01)
        data.extend(pack_int32(int(e.pos.x)))
        e.last_state['x'] = int(e.pos.x)

    # print(f"{e.last_state['y']}:{e.pos.y}")
    if e.last_state.get('y') != e.pos.y:
        # print("Added y")
        data.append(0x02)
        data.extend(pack_int32(int(e.pos.y)))
        e.last_state['y'] = int(e.pos.y)

    # print(f"{e.last_state['state']}:{e.state}")
    if e.last_state.get('state') != e.state:
        # print("Added state")
        data.append(0x03)
        data.extend(pack_uint8(e.state))
        e.last_state['state'] = int(e.state)

    # print(f"{e.last_state['direction']}:{e.direction}")
    if e.last_state.get('direction') != e.direction:
        # print("Added direction")
        data.append(0x04)
        data.extend(pack_uint8(e.direction))
        e.last_state['direction'] = int(e.direction)

    return bytes(data)

def serialise_entities(es: List[Entity]) -> bytes:
    data = bytearray()
    l = len(es)
    num_entities = struct.pack('B', l)
    data.extend(num_entities)

    for e in es:
        server_id = struct.pack('!H', e.in_server_id)
        data.append(0x00)
        data.extend(server_id)
        if e.last_state.get('x') != e.pos.x:
            data.append(0x01)
            data.extend(pack_int32(int(e.pos.x)))
            e.last_state['x'] = int(e.pos.x)

        if e.last_state.get('y') != e.pos.y:
            data.append(0x02)
            data.extend(pack_int32(int(e.pos.y)))
            e.last_state['y'] = int(e.pos.y)

        if e.last_state.get('state') != e.state:
            data.append(0x03)
            data.extend(pack_uint8(e.state))
            e.last_state['state'] = int(e.state)

        if e.last_state.get('direction') != e.direction:
            data.append(0x04)
            data.extend(pack_uint8(e.direction))
            e.last_state['direction'] = int(e.direction)

    return bytes(data)

"""
# get number of entities as uint8
# then entities
eg:
    0x05                        # indicates 5 entities
    0x00 0x00 0x44              # indicates in server id of 68
    0x01 0x00 0x00 0x00 0xa5    # indicats position x (0x01) is 165 (0xa0)
    0x00 0x00 0x45              # indicates in server id of 69
    0x01 0x00 0x00 0x00 0x00    # indicats position x (0x01) is 0
    0x02 0x00 0x00 0x00 0x00    # indicats position y (0x02) is 0
    ...
"""

def deserialise(data: bytes) -> Entity:
    # only one entity including in server id
    i = 0
    # print(data)
    if data[i] != 0x00:
        raise Exception("Invalid data")
    i += 1
    in_server_id, i = unpack_uint16(data, i)
    entity = Entity()
    entity.in_server_id = in_server_id  
    while i < len(data):
        field_code = data[i]
        i += 1 # to move to data for unpack_... to have data iteslf without field code
        # print(field_code)
        if field_code == 0x01:
            # print(f"	x: {data[i:i+4]}")
            entity.pos.x, i = unpack_int32(data, i)
        elif field_code == 0x02:
            # print(f"	y: {data[i:i+4]}")
            entity.pos.y, i = unpack_int32(data, i)
        elif field_code == 0x03:
            # print(f"	state: {data[i:i+1]}")
            entity.state, i = unpack_uint8(data, i)
        elif field_code == 0x04:
            # print(f"dirrection: {data[i:i+1]}")
            entity.direction, i = unpack_uint8(data, i)
        # for multiple entities implementation
        elif field_code == 0x00:
            break
        else:
            raise Exception("Trailing invalid data")
    return entity


def deserialise_entities(data: bytes) -> List[Entity]:
    entities = []
    i = 0

    # !!! Can't use deserialise(data) because i should increment and it would
    #     be redundent to add i to deserialise

    # Get the number of entities
    num_entities, i = unpack_uint8(data, i)
    # sould be one. point at field code 0x00 server id
    # print("Deserialising entities: ",num_entities)

    while i < len(data):
        # Read the field code for server ID
        field_code = data[i]
        i += 1 # should point at field code
        # print(f"Field code: {field_code}")


        if field_code == 0x00:
            server_id, i = unpack_uint16(data, i)

            entity = Entity()
            entity.in_server_id = server_id
            entity.pos = pygame.Vector2(0,0)
            # print(f"For entity with isid: {server_id}:")

            while i < len(data):
                field_code = data[i]
                # print(f"Field code: {field_code}")
                i += 1 # point at data
                if field_code == 0x00:
                    # If a new entity starts, break the loop
                    i -= 1
                    break
                elif field_code == 0x01:
                    # print(f"\tx: {data[i:i+4]}")
                    entity.pos.x, i = unpack_int32(data, i)
                elif field_code == 0x02:
                    # print(f"\ty: {data[i:i+4]}")
                    entity.pos.y, i = unpack_int32(data, i)
                elif field_code == 0x03:
                    # print(f"\tstate: {data[i:i+1]}")
                    entity.state, i = unpack_uint8(data, i)
                elif field_code == 0x04:
                    # print(f"\tdirrection: {data[i:i+1]}")
                    entity.direction, i = unpack_uint8(data, i)
            entities.append(entity)

    return entities


if __name__ == "__main__":
    # test multiple entities
    def print_entities(es: List[Entity], msg:str=""):
        indent=""
        if msg != "":
            print(msg)
            indent = "\t"
        for e in es:
            print(f"{indent}Entity: {e.in_server_id}@{e.pos.x: >5}:{e.pos.y: >5}(s:{e.state};d:{e.direction}")
    entities = [
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
    ]
    print_entities(entities, "Before serialisation")
    serialised_entities = serialise_entities(entities)
    print("serialised_entities")
    deserialised_entities = deserialise_entities(serialised_entities)
    print_entities(deserialised_entities, "Deserialised entities:")
    
    # Test individual entities
    serialised_entities_array = []
    entities = [
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
        Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3)),
    ]
    print_entities(entities, "before serialising individual entities")
    for e in entities:
        b = bytearray()
        b.append(0x00)
        b.extend(pack_uint16(e.in_server_id))
        b.extend(serialise(e))
        serialised_entities_array.append(b)

    print("Serialised individual entities")
    deserialised_entities_array = []
    for e in serialised_entities_array:
        deserialised_entities_array.append(deserialise(e))
    print_entities(deserialised_entities_array, "Deserialised individual entities")
    """
    e = Entity(isid=randint(0, 1234),x=randint(-10000,10000),y=randint(-10000,10000),state=randint(0,3),direction=randint(0,3))
    print(f"Entity: {e.in_server_id}@{e.pos.x: >5}:{e.pos.y: >5}(s:{e.state};d:{e.direction}")
    serialised_e = bytearray()
    serialised_e.append(0x00)
    serialised_e.extend(pack_uint16(e.in_server_id))
    serialised_e.extend(serialise(e))
    deserialised_e = deserialise(serialised_e)
    print_entities([deserialised_e,],"Deserailised e:")
    """
    

File: Server/src/serialisation.go
package src

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"math/rand"
)

type Change[T any] struct  {
    Changed         bool
    Value           T
}

type SerialisationEntityStruct struct {
    InServerID      uint16
    X               Change[int32]
    Y               Change[int32]
    State           Change[uint8]
    Direction       Change[uint8]
}

func packInt32(value int32) []byte {
	buf := new(bytes.Buffer)
	err := binary.Write(buf, binary.BigEndian, value)
	if err != nil {
		fmt.Println("binary.Write(packInt32) failed:", err)
	}
	return buf.Bytes()
}

func packUint16(value uint16) []byte {
	buf := new(bytes.Buffer)
	err := binary.Write(buf, binary.BigEndian, value)
	if err != nil {
		fmt.Println("binary.Write(packInt16) failed:", err)
	}
	return buf.Bytes()
}

func packUint8(value uint8) []byte {
	buf := new(bytes.Buffer)
	err := binary.Write(buf, binary.BigEndian, value)
	if err != nil {
		fmt.Println("binary.Write(packInt8) failed:", err)
	}
	return buf.Bytes()
}

// Function to unpack an int32 value from a byte slice with a given offset
func unpackInt32(data []byte, offset int) (int32, int, error) {
	var value int32

	buf := bytes.NewReader(data[offset:])
	err := binary.Read(buf, binary.BigEndian, &value)
	if err != nil {
		return 0, offset, fmt.Errorf("binary.Read(unpackInt32) failed: %w", err)

	}
	return value, offset + 4, nil

}


// Function to unpack a uint16 value from a byte slice with a given offset
func unpackUint16(data []byte, offset int) (uint16, int, error) {

	var value uint16
	buf := bytes.NewReader(data[offset:])
	err := binary.Read(buf, binary.BigEndian, &value)
	if err != nil {
		return 0, offset, fmt.Errorf("binary.Read(unpackInt16) failed: %w", err)
	}
	return value, offset + 2, nil
}

func unpackUint8(data []byte, offset int) (uint8, int, error) {
	var value uint8
	buf := bytes.NewReader(data[offset:])
	err := binary.Read(buf, binary.BigEndian, &value)
	if err != nil {
		fmt.Println("binary.Read failed:", err)
		return 0, offset, fmt.Errorf("binary.Read(unpackInt8) failed: %w", err)
	}
	return value, offset + 1, nil
}

// Returns empty array if everything fails or entity did not change
// Does not include inServerID
func serialise(e *SerialisationEntityStruct) []byte {
	data := new(bytes.Buffer) 

	if e.X.Changed {
		data.WriteByte(0x01)
		data.Write(packInt32(int32(e.X.Value)))
	}

	if e.Y.Changed {
		data.WriteByte(0x02)
		data.Write(packInt32(int32(e.Y.Value)))
	}

	if e.State.Changed {
		data.WriteByte(0x03)
		data.Write(packUint8(e.State.Value))
	}

	if e.Direction.Changed {
		data.WriteByte(0x04)
		data.Write(packUint8(e.Direction.Value))
	}

	return data.Bytes()
}
// returns an empty array if it failes completly
// doesn't include entities that fail to serialise
func SerialiseEntities(es []*SerialisationEntityStruct) []byte {
	data := new(bytes.Buffer)
    entitiesLength := len(es)
    length := packUint8(uint8(entitiesLength))
    data.Write(length)

    totalEntitiesWritten := 0
    for _,e := range es {
        entityData := new(bytes.Buffer)

        inServerID := packUint16(e.InServerID)
        // one byte
        entityData.WriteByte(0x00)
        // 2 bytes
        entityData.Write(inServerID)

        if entityData.Len() < 3 {
            fmt.Println("Failed to serialise Entity inServerID")
            continue
        }

        serialisedEntityData := serialise(e)
        if len(serialisedEntityData) <= 0 {
            fmt.Println("Failed to serialise Entity/Entity might not have changed")
            continue
        }
        entityData.Write(serialisedEntityData)
        // Should be greater than three / 1 for 0x00, one for in server id and 1+ from entity serialisation
        if entityData.Len() <= 3 {
            fmt.Println("Failed to serialise Entity")
            continue
        } else {
            data.Write(entityData.Bytes())
            totalEntitiesWritten ++
        }
    }
    // fmt.Printf("Serialised a total of %v out of %v entities. len data: %v\n", totalEntitiesWritten, length, data.Len())
    return data.Bytes()
}

// Returns 
func Deserialise(data []byte) (*SerialisationEntityStruct, error) {
    // ignoring health for now
    e := SerialisationEntityStruct{
        0,
        Change[int32]   {false, 0},
        Change[int32]   {false, 0},
        Change[uint8]   {false, 0},
        Change[uint8]   {false, 0},
    }
    i := 0
    if data[i] != 0x00 {
        fmt.Printf("Invalid fieldCode %v. first byte should be 0x00 followed by inServerID\n", string(data[i]))
        return nil, fmt.Errorf("Invalid fieldCode %v. first byte should be 0x00 followed by inServerID\n", string(data[i]))
    }
    i++

    var err error
    e.InServerID, i, err = unpackUint16(data, i)
    if err != nil {
        fmt.Printf("Fialed to deserialise inServerID/Invalid inServerID: %v.")
        return nil,fmt.Errorf("Fialed to deserialise inServerID/Invalid inServerID: %v.")
    }

    for i < len(data) {
        fieldCode := data[i]
        i++
        err = nil

        switch fieldCode {
        case 0x01:
            e.X.Value, i, err = unpackInt32(data, i)
            if err != nil {
                fmt.Println("Failed to deserialise Entity x position:", err)
            } else {
                e.X.Changed = true
            }
        case 0x02:
            e.Y.Value, i, err = unpackInt32(data, i)
            if err != nil {
                fmt.Println("Failed to deserialise Entity y position:", err)
            } else {
                e.Y.Changed = true
            }
        case 0x03:
            e.State.Value, i, err = unpackUint8(data, i)
            if err != nil {
                fmt.Println("Failed to deserialise Entity state:", err)
            } else {
                e.State.Changed = true
            }
        case 0x04:
            e.Direction.Value, i, err = unpackUint8(data, i)
            if err != nil {
                fmt.Println("Failed to deserialise Entity direction:", err)
            } else {
                e.Direction.Changed = true
            }
        default:
            i--
            fmt.Println("Invalid Field Code at: %v", i)
                return nil, fmt.Errorf(
                    "Invalid Field Code at: %v. %v", i,
                    "This will cause the rest of the data to not be valid.")
        }
    }
    return &e, nil
}

func DeserialiseEntityFrom(data []byte, i int) (*SerialisationEntityStruct, int, error) {
    // ignoring health for now
    e := SerialisationEntityStruct{
        0,
        Change[int32]   {false, 0},
        Change[int32]   {false, 0},
        Change[uint8]   {false, 0},
        Change[uint8]   {false, 0},
    }

    if i >= len(data) {
        return nil, i, fmt.Errorf("data index out of range")
    }

    fieldCode := data[i]
    i++
    if fieldCode != 0x00 {
        return nil, i, fmt.Errorf("invalid field code. expected 0x00, got %x", fieldCode)
    }

    var err error
    e.InServerID, i, err = unpackUint16(data, i)
    if err != nil {
        return nil, i, fmt.Errorf("failed to deserialise InServerID: %w", err)
    }

loop:
    for i < len(data) {
        fieldCode = data[i]
        i++
        switch fieldCode {
        case 0x01:
            e.X.Value, i, err = unpackInt32(data, i)
            if err != nil {
                fmt.Println("Failed to unpack X position:", err)
            } else {
                e.X.Changed = true
            }
        case 0x02:
            e.Y.Value, i, err = unpackInt32(data, i)
            if err != nil {
                fmt.Println("Failed to unpack Y position:", err)
            } else {
                e.Y.Changed = true
            }
        case 0x03:
            e.State.Value, i, err = unpackUint8(data, i)
            if err != nil {
                fmt.Println("Failed to unpack State:", err)
            } else {
                e.State.Changed = true
            }
        case 0x04:
            e.Direction.Value, i, err = unpackUint8(data, i)

            if err != nil {

                fmt.Println("Failed to unpack Direction:", err)
            } else {
                e.Direction.Changed = true
            }
        case 0x00:
            i--
            break loop
        default:
            return &e, i, fmt.Errorf("unknown field code: %x", fieldCode)

        }
    }
    
    return &e, i, nil
}

func DeserialiseEntities(data []byte) ([]*SerialisationEntityStruct, error) {
    i := 0
    entitiesLength, i, err := unpackUint8(data, i)
    if err != nil{
        fmt.Printf("Failed to decode length: %v", err)
        return nil, fmt.Errorf("Failed to decode length")
    }
    
    entities := make([]*SerialisationEntityStruct, entitiesLength)
    entitiesIndex := 0

    for i < len(data) && entitiesIndex < int(entitiesLength) {
        fieldCode := data[i]
        if fieldCode != 0x00 {
            fmt.Println("invalid fieldCode")
            return nil, fmt.Errorf("invalid fieldCode")
        }
        var entity *SerialisationEntityStruct
        entity, i, err = DeserialiseEntityFrom(data, i)
        if err != nil {
            fmt.Println("Error deserialising entity:", err)
            return entities[:entitiesIndex], fmt.Errorf(
                "Failed to deserialise entity%v. Returning successful entities. %v",
                err, "Data read from here on might be invalid. Therefore skipping.",
                )
        }
        // fmt.Printf("Deserialised entity: %v\n", entity.InServerID)
        entities[entitiesIndex] = entity
        entitiesIndex++
    }
    // fmt.Printf("Deserialised %v out of %v entities:", entitiesIndex, entitiesLength)
    return entities[:entitiesIndex], nil
}



// test main
func NewSerialisationEntityStruct(isid uint16, x int32, y int32, state uint8, dir uint8) *SerialisationEntityStruct {
    e := &SerialisationEntityStruct{}
    e.InServerID = isid
    e.X = Change[int32] {true, x}
    e.Y = Change[int32] {true, y}
    e.State = Change[uint8] {true, state}
    e.Direction = Change[uint8] {true, dir}
    return e
}
func NSES(isid uint16, x int32, y int32, state uint8, dir uint8) *SerialisationEntityStruct {
    e := &SerialisationEntityStruct{}
    e.InServerID = isid
    e.X = Change[int32] {true, x}
    e.Y = Change[int32] {true, y}
    e.State = Change[uint8] {true, state}
    e.Direction = Change[uint8] {true, dir}
    return e
}

func PrintSESs(es []*SerialisationEntityStruct, msg string) {
    if msg == "" {
        for _, e := range es {
            fmt.Printf("Entity: %v@%v:%v(s:%v;d:%v)\n", e.InServerID, e.X.Value, e.Y.Value, e.State.Value, e.Direction.Value)
        }
        return
    }
    fmt.Println(msg)
    for _, e := range es {
        fmt.Printf("\tEntity: %v@%v:%v(s:%v;d:%v)\n", e.InServerID, e.X.Value, e.Y.Value, e.State.Value, e.Direction.Value)
    }
}


// randint generates a random integer between a and b (inclusive)
func randint(a, b int) int {
    return rand.Intn(b-a+1) + a
}

func TestSingleEntity() {
    errors := make([]error, 0)
    // Creation
    testEntity := NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40)))
    PrintSESs([]*SerialisationEntityStruct{testEntity}, "testEntity:")

    // Serialisation
    serialisedTestEntity := new(bytes.Buffer) 
    serialisedTestEntity.WriteByte(0x00)
    serialisedTestEntity.Write(packUint16(testEntity.InServerID))
    serialisedTestEntity.Write(serialise(testEntity))
        // to test if new entity
    serialisedTestEntity.WriteByte(0x00)

    deserialisedTestEntity, _, err := DeserialiseEntityFrom(serialisedTestEntity.Bytes(), 0)
    if err != nil {
        fmt.Printf("Failed to deserialise testEntity: %v\n", err)
        errors = append(errors, fmt.Errorf("Failed to deserialise testEntity: %v\n", err))
    } else {
    // Test
        // if success
        if testEntity.X.Value == deserialisedTestEntity.X.Value &&
        testEntity.Y.Value == deserialisedTestEntity.Y.Value &&
        testEntity.State.Value == deserialisedTestEntity.State.Value &&
        testEntity.Direction.Value == deserialisedTestEntity.Direction.Value {
            fmt.Println("Entities are the same. success.")
        } else {
            PrintSESs([]*SerialisationEntityStruct{deserialisedTestEntity},
            "Test failed. Deserialised testEntity:")
        }
    }
}

func TestMultipleEntities() {
    errors := make([]error, 0)
    // Creation
    testEntities := []*SerialisationEntityStruct{
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
        NSES(uint16(randint(0,1234)), int32(randint(0,12345)), int32(randint(0,12345)),uint8(randint(0,40)), uint8(randint(0,40))),
    }
    PrintSESs(testEntities, "testEntities:")
    // Serialisation
    serialisedTestEntities := SerialiseEntities(testEntities)
    deserialisedTestEntities, err := DeserialiseEntities(serialisedTestEntities)
    if err != nil {
        fmt.Printf("Failed to deserialise testEntities: %v\n", err)
        errors = append(errors, fmt.Errorf("Failed to deserialise testEntities: %v\n", err))
    }
    // Test
    if len(deserialisedTestEntities) != len(testEntities) {
        fmt.Printf("Entities length is not the same: testEntities: %v; deserialisedTestEntities: %v",
            len(testEntities), len(deserialisedTestEntities))
    } else {
        // if success
        errs := 0
        for i := 0; i < len(deserialisedTestEntities); i++ {
            testEntity := testEntities[i]
            deserialisedTestEntity := deserialisedTestEntities[i]
            if      testEntity.X.Value != deserialisedTestEntity.X.Value     ||
                    testEntity.Y.Value != deserialisedTestEntity.Y.Value     ||
                testEntity.State.Value != deserialisedTestEntity.State.Value ||
            testEntity.Direction.Value != deserialisedTestEntity.Direction.Value {
                fmt.Println("Entities at %v aren't the same. %v:%v\n", i, testEntities[i], deserialisedTestEntities[i])
                errs += 1
            }
        }
        if errs != 0 {
            fmt.Printf("%v entities aren't the same.\n", errs)
            PrintSESs(deserialisedTestEntities, "DeserialiseEntities:")
        } else {
            fmt.Println("All entities are the same. success")
        }
    }
}
